### ISSUE TITLE
[Bug]: 

### ISSUE THREAD

**Issue by Bearsaerker** [run-llama/llama_index#10191]
### Bug Description

I have a peculiar problem. As of llama index version 0.9.22 ollama is not able to produce text for me. I downgraded and tested almost all versions from 0.9.1 up to 0.9.21. 0.9.21 is the last version in which ollama is able to produce text with llama index.
I have it integrated as " llm = Ollama(model="Solar", temperature=0.2)" Adding request_timeout doesn't change a thing unfortunately

### Version

0.9.33

### Steps to Reproduce

Downgrading to below 0.9.22 eliminates the issue, above, ollama can't generate text.

**Comment by logan-markewich** [comment]
Is there some error? Any other details?

**Comment by Bearsaerker** [comment]
Unfortunately not, it just does nothing. It retrieves nodes and the nodes are also printed if I use print(source.node.get_text()). At first I thought it may be a problem with my system and I completely reinstalled fedora, but nothing changed.

**Comment by logan-markewich** [comment]
I'm fairly sure I've tested this recently and it worked fine... let me try again

**Comment by Bearsaerker** [comment]
So I did the following: 
    print("Constructing query engine...\n")
    query_engine = index.as_chat_engine(chat_mode="context", similarity_top_k=40, node_postprocessors=[cohere_rerank], streaming=True, system_prompt=("My way too long prompt.")) 

    response = query_engine.chat("What is 9 + 10?")
    
    print(response)
    
    And I get the following error:
    
    Traceback (most recent call last):
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_exceptions.py", line 10, in map_exceptions
    yield
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_backends/sync.py", line 126, in read
    return self._sock.recv(max_bytes)
TimeoutError: timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_transports/default.py", line 66, in map_httpcore_exceptions
    yield
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_transports/default.py", line 228, in handle_request
    resp = self._pool.handle_request(req)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py", line 268, in handle_request
    raise exc
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py", line 251, in handle_request
    response = connection.handle_request(request)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/connection.py", line 103, in handle_request
    return self._connection.handle_request(request)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 133, in handle_request
    raise exc
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 111, in handle_request
    ) = self._receive_response_headers(**kwargs)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 176, in _receive_response_headers
    event = self._receive_event(timeout=timeout)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 212, in _receive_event
    data = self._network_stream.read(
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_backends/sync.py", line 124, in read
    with map_exceptions(exc_map):
  File "/home/luis/.conda/envs/vllm/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ReadTimeout: timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 535, in _run_script
    exec(code, module.__dict__)
  File "/home/luis/Dokumente/ChatWithDoc English/Streamlit.py", line 130, in <module>
    query_engine = setup_ollama_and_weaviate()
  File "/home/luis/Dokumente/ChatWithDoc English/Streamlit.py", line 124, in setup_ollama_and_weaviate
    response = query_engine.chat("What is 9 + 10?")
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/llama_index/callbacks/utils.py", line 41, in wrapper
    return func(self, *args, **kwargs)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/llama_index/chat_engine/context.py", line 163, in chat
    chat_response = self._llm.chat(all_messages)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/llama_index/llms/base.py", line 97, in wrapped_llm_chat
    f_return_val = f(_self, messages, **kwargs)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/llama_index/llms/ollama.py", line 102, in chat
    response = client.post(
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_client.py", line 1132, in post
    return self.request(
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_client.py", line 814, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_client.py", line 901, in send
    response = self._send_handling_auth(
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_client.py", line 929, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_client.py", line 966, in _send_handling_redirects
    response = self._send_single_request(request)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_client.py", line 1002, in _send_single_request
    response = transport.handle_request(request)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_transports/default.py", line 227, in handle_request
    with map_httpcore_exceptions():
  File "/home/luis/.conda/envs/vllm/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/luis/.conda/envs/vllm/lib/python3.10/site-packages/httpx/_transports/default.py", line 83, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadTimeout: timed out

**Comment by logan-markewich** [comment]
OK that makes more sense, a timeout error.

And you tried setting `request_timeout` to some absurdly large value?

**Comment by logan-markewich** [comment]
You can see request_timeout is used consistently in the code
https://github.com/run-llama/llama_index/blob/e34c379f1a1c46b775580ae8678886a44cf775ca/llama_index/llms/ollama.py#L101

**Comment by Bearsaerker** [comment]
If I do that, it will generate a response. But somehow the overarching function is still broken. I use it inside a streamlit integration and with llama index 0.9.21 it is absolutely no problem

**Comment by logan-markewich** [comment]
"It will generate a response. But somehow the overarching function is still broken." -- but what does that mean?

**Comment by logan-markewich** [comment]
Trying something somewhat similar, it works fine for me (and this is running on CPU, so quite slow, as expected)

```python
from llama_index.llms import Ollama
from llama_index import ServiceContext, SimpleDirectoryReader, VectorStoreIndex

llm = Ollama(model="starling-lm:latest", request_timeout=3000)
ctx = ServiceContext.from_defaults(llm=llm, embed_model="local")

documents = SimpleDirectoryReader("./docs/examples/data/paul_graham").load_data()
index = VectorStoreIndex.from_documents(documents, service_context=ctx)

chat_engine = index.as_chat_engine(chat_mode="context")
resp = chat_engine.chat("What did the author do growing up?")
print(str(resp))
```

**Comment by Bearsaerker** [comment]
```python

def setup_ollama_and_weaviate():
    with open('/home/luis/Dokumente/ChatWithDoc English/config.yml', 'r', encoding='utf8') as ymlfile:
        cfg = box.Box(yaml.safe_load(ymlfile))
    
    print("Connecting to Weaviate")
    client = weaviate.Client(cfg.WEAVIATE_URL)

    print("Loading Ollama...")
    llm = Ollama(model="FusionNetTruthQ4", temperature=0.2)

    print("Loading embedding model...")
    embeddings = load_embedding_model(model_name='jinaai/jina-embeddings-v2-base-en')

    print("Building index...")
    index = build_index(cfg.CHUNK_SIZE, cfg.CHUNK_OVERLAP, llm, embeddings, client, cfg.INDEX_NAME)
   
    cohere_rerank = CohereRerank(api_key="bW8KGD0GqZdLvx4QcRg6OQfROvkMuhMAZd0CtIP3", top_n=6)
    
    print("Constructing query engine...\n")
    query_engine = index.as_chat_engine(chat_mode="condense_plus_context", similarity_top_k=40, node_postprocessors=[cohere_rerank], streaming=True, context_prompt=("My very long prompt.")) 
  
    return query_engine
    
    query_engine = setup_ollama_and_weaviate()

# Define an enum for roles
class Role(Enum):
    USER = "user"
    ASSISTANT = "assistant"

# Update the ChatMessage class
class ChatMessage:
    def __init__(self, role, content, additional_kwargs=None):
        if additional_kwargs is None:
            additional_kwargs = {}
        self.role = Role(role)
        self.content = content
        self.additional_kwargs = additional_kwargs

# Chat input processing

if prompt := st.chat_input("What is up?"):
    st.session_state.messages.append({"role": "user", "content": prompt})
    with st.chat_message("user", avatar="/home/luis/Dokumente/ChatWithDoc English/capri.jpg"):
        st.markdown(prompt)

    with st.chat_message("assistant", avatar="/home/luis/Dokumente/ChatWithDoc English/ollama.png"):
        message_placeholder = st.empty()
        full_response = ""

        # Convert st.session_state.messages to the expected format
        chat_history = [ChatMessage(m["role"], m["content"]) for m in st.session_state.messages]

        # Use the 'chat' method for generating responses
        streaming_response = query_engine.stream_chat(prompt, chat_history=chat_history)
        
        for source in streaming_response.source_nodes:
            # This prints the text of the node
            print(source.node.get_text())

            # Check if 'title' exists in the metadata of the node and print it
            if 'file_path' in source.node.metadata:
                print("\033[92m" + source.node.metadata['file_path'] + "\033[0m")  # This prints the title if it exists in metadata
                print("\n")

        for token in streaming_response.response_gen:
            full_response += token
            message_placeholder.markdown(full_response)
            
        st.session_state.messages.append({"role": "assistant", "content": full_response})
```
    
If I put the print command before the "return query_engine" it generates an output. If I put it afterwards, it won't and streamlit won't process my prompt. As a little fix I just copied the old ollama integration into the newer llama index and that seems to work fine.

**Comment by Bearsaerker** [comment]
```python
def setup_ollama_and_weaviate():
    with open('/home/luis/Dokumente/ChatWithDoc English/config.yml', 'r', encoding='utf8') as ymlfile:
        cfg = box.Box(yaml.safe_load(ymlfile))
    
    print("Connecting to Weaviate")
    client = weaviate.Client(cfg.WEAVIATE_URL)

    print("Loading Ollama...")
    llm = Ollama(model="FusionNetTruthQ4", temperature=0.2)

    print("Loading embedding model...")
    embeddings = load_embedding_model(model_name='jinaai/jina-embeddings-v2-base-en')

    print("Building index...")
    index = build_index(cfg.CHUNK_SIZE, cfg.CHUNK_OVERLAP, llm, embeddings, client, cfg.INDEX_NAME)
   
    cohere_rerank = CohereRerank(api_key=api_key, top_n=6)
    
    print("Constructing query engine...\n")
    query_engine = index.as_chat_engine(chat_mode="condense_plus_context", similarity_top_k=40, node_postprocessors=[cohere_rerank], streaming=True, context_prompt=("My very long prompt.")) 
  
    return query_engine


query_engine = setup_ollama_and_weaviate()

# Define an enum for roles
class Role(Enum):
    USER = "user"
    ASSISTANT = "assistant"

# Update the ChatMessage class
class ChatMessage:
    def __init__(self, role, content, additional_kwargs=None):
        if additional_kwargs is None:
            additional_kwargs = {}
        self.role = Role(role)
        self.content = content
        self.additional_kwargs = additional_kwargs


# Chat input processing
# Chat input processing
if prompt := st.chat_input("What is up?"):
    st.session_state.messages.append({"role": "user", "content": prompt})
    with st.chat_message("user", avatar="/home/luis/Dokumente/ChatWithDoc English/capri.jpg"):
        st.markdown(prompt)

    with st.chat_message("assistant", avatar="/home/luis/Dokumente/ChatWithDoc English/ollama.png"):
        message_placeholder = st.empty()
        full_response = ""

        # Convert st.session_state.messages to the expected format
        chat_history = [ChatMessage(m["role"], m["content"]) for m in st.session_state.messages]

        # Use the 'chat' method for generating responses
        streaming_response = query_engine.stream_chat(prompt, chat_history=chat_history)
        
        for source in streaming_response.source_nodes:
            # This prints the text of the node
            print(source.node.get_text())

            # Check if 'title' exists in the metadata of the node and print it
            if 'file_path' in source.node.metadata:
                print("\033[92m" + source.node.metadata['file_path'] + "\033[0m")  # This prints the title if it exists in metadata
                print("\n")

        for token in streaming_response.response_gen:
            full_response += token
            message_placeholder.markdown(full_response)
            
        st.session_state.messages.append({"role": "assistant", "content": full_response})
 ```

If I put the print command before the "return query_engine" it generates an output. If I put it afterwards, it won't and streamlit won't process my prompt. As a little fix I just copied the old ollama integration into the newer llama index and that seems to work fine.

**Comment by logan-markewich** [comment]
@Bearsaerker I've narrowed this down a bit.... seems to be related to streaming, as normal `chat()` works fine.

Not 100% sure how to further debug this, maybe it's related to `httpx` and how it handles streaming?

**Comment by logan-markewich** [comment]
WOW this was a sneaky bug. Some non-serializable thing was only `sometimes` making its way into requests? And the error was only sometimes getting raised? 

Very weird. Will have a PR in a sec, but I got your sample code to actually work
